### PR TITLE
TransT tracker integration

### PR DESCRIPTION
**Pull request by dschoerk**
PR to integrate the single object tracker TransT as an AI tool into CVAT.

also see here: https://github.com/cvat-ai/cvat-opencv/issues/14

### CODE REVIEW

File: cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/tools-control.tsx [hunk 1]
```diff
@@ -756,7 +756,7 @@ export class ToolsControlComponent extends React.PureComponent<Props, State> {
                         });
                         // eslint-disable-next-line no-await-in-loop
                         const response = await core.lambda.call(jobInstance.taskId, tracker, {
-                            frame: frame,
+                            frame: frame ,
```
Comment (dschoerk) [inline cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/tools-control.tsx:759]: to avoid the linter issue

**Comment by kirill-sizov** [comment]
Resolved #4768

**Comment by tangy5** [comment]
Hi @dschoerk , thank for this really cool integration. This is very inspiring! 
Can I ask how you write the code for annotating video in CVAT? I'm trying to implement a different task with endoscopy videos, I saw your video shows the CVAT processed each frame at a time until all frames are tracked/inference. 

How are the images fed/transformed? In meantime, how are those requests kept posting to nutlio server until all frames(images) are done? 

Thank you in advance and it would be super appreciated if there sample code for this video integration! Thanks!

**Comment by dschoerk** [comment]
> Hi @dschoerk , thank for this really cool integration. This is very inspiring! Can I ask how you write the code for annotating video (MP4) in CVAT? I'm trying to implement a different task with endoscopy videos, I saw your video shows the CVAT processed each frame at a time until all frames are tracked/inference.
> 
> How are the images fed/transformed? In meantime, how are those requests kept posting to nutlio server until all frames(images) are done?
> 
> Thank you in advance and it would be super appreciated if there sample code for this video integration! Thanks!

at this time CVAT only supports to track one step at a time. a bounding box (seed) is drawn on the initially frame, and each time you press the "f" key to step to the next frame the objects are tracked. afaik there is no functionality at the moment to track multiple frames at once. in the mentioned video i just keep pressing the "f" key. i hope this answers your question.

**Comment by tangy5** [comment]
> > Hi @dschoerk , thank for this really cool integration. This is very inspiring! Can I ask how you write the code for annotating video (MP4) in CVAT? I'm trying to implement a different task with endoscopy videos, I saw your video shows the CVAT processed each frame at a time until all frames are tracked/inference.
> > How are the images fed/transformed? In meantime, how are those requests kept posting to nutlio server until all frames(images) are done?
> > Thank you in advance and it would be super appreciated if there sample code for this video integration! Thanks!
> 
> at this time CVAT only supports to track one step at a time. a bounding box (seed) is drawn on the initially frame, and each time you press the "f" key to step to the next frame the objects are tracked. afaik there is no functionality at the moment to track multiple frames at once. in the mentioned video i just keep pressing the "f" key. i hope this answers your question.

Thank you, this is very helpful. Helped me understand CVAT much better. One step further, do you think there is can be functionality to automate the "next step" and "prediction" with each frame without pressing "f", until all frames are done?
Thank you again on the reply! Very appreciated. CVAT is a great tool.

**Comment by tangy5** [comment]
> > > Hi @dschoerk , thank for this really cool integration. This is very inspiring! Can I ask how you write the code for annotating video (MP4) in CVAT? I'm trying to implement a different task with endoscopy videos, I saw your video shows the CVAT processed each frame at a time until all frames are tracked/inference.
> > > How are the images fed/transformed? In meantime, how are those requests kept posting to nutlio server until all frames(images) are done?
> > > Thank you in advance and it would be super appreciated if there sample code for this video integration! Thanks!
> > 
> > 
> > at this time CVAT only supports to track one step at a time. a bounding box (seed) is drawn on the initially frame, and each time you press the "f" key to step to the next frame the objects are tracked. afaik there is no functionality at the moment to track multiple frames at once. in the mentioned video i just keep pressing the "f" key. i hope this answers your question.
> 
> Thank you, this is very helpful. Helped me understand CVAT much better. One step further, do you think there can be a functionality to automate the "next step" and "prediction" with each frame without pressing "f", until all frames are done? Thank you again on the reply! Very appreciated. CVAT is a great tool.

**Comment by dschoerk** [comment]
> > > Hi @dschoerk , thank for this really cool integration. This is very inspiring! Can I ask how you write the code for annotating video (MP4) in CVAT? I'm trying to implement a different task with endoscopy videos, I saw your video shows the CVAT processed each frame at a time until all frames are tracked/inference.
> > > How are the images fed/transformed? In meantime, how are those requests kept posting to nutlio server until all frames(images) are done?
> > > Thank you in advance and it would be super appreciated if there sample code for this video integration! Thanks!
> > 
> > 
> > at this time CVAT only supports to track one step at a time. a bounding box (seed) is drawn on the initially frame, and each time you press the "f" key to step to the next frame the objects are tracked. afaik there is no functionality at the moment to track multiple frames at once. in the mentioned video i just keep pressing the "f" key. i hope this answers your question.
> 
> Thank you, this is very helpful. Helped me understand CVAT much better. One step further, do you think there is can be functionality to automate the "next step" and "prediction" with each frame without pressing "f", until all frames are done? Thank you again on the reply! Very appreciated. CVAT is a great tool.

to extremely simplify things: AI tools are integrated as serverless functions i.e. they get called via a rest interface on the nuclio platform like some webservice. an image is sent from cvat to the service and it responds with the tracked location and state of the tracker. within this PR i have implemented such a service. from the perspective of implementation this is great, because of its simplicity - performance when tracking over multiple frames is not amazing with this approach. 

to implement what you're looking for is not trivial. the simplest i can imagine is to call the tracking service repeatedly until the required amount of frames is tracked. BUT this is not very performant. each frame is sent in a separate http request to the service and it requires n requests to tracking n frames. a better solution would be to have a service that's capable of tracking multiple frames, keep its state internally and doesn't require the images to be sent in the request but rather access them from a docker mount. all of that would reduce flexibility but increase performance.

**Comment by tangy5** [comment]
> > > > Hi @dschoerk , thank for this really cool integration. This is very inspiring! Can I ask how you write the code for annotating video (MP4) in CVAT? I'm trying to implement a different task with endoscopy videos, I saw your video shows the CVAT processed each frame at a time until all frames are tracked/inference.
> > > > How are the images fed/transformed? In meantime, how are those requests kept posting to nutlio server until all frames(images) are done?
> > > > Thank you in advance and it would be super appreciated if there sample code for this video integration! Thanks!
> > > 
> > > 
> > > at this time CVAT only supports to track one step at a time. a bounding box (seed) is drawn on the initially frame, and each time you press the "f" key to step to the next frame the objects are tracked. afaik there is no functionality at the moment to track multiple frames at once. in the mentioned video i just keep pressing the "f" key. i hope this answers your question.
> > 
> > 
> > Thank you, this is very helpful. Helped me understand CVAT much better. One step further, do you think there is can be functionality to automate the "next step" and "prediction" with each frame without pressing "f", until all frames are done? Thank you again on the reply! Very appreciated. CVAT is a great tool.
> 
> to extremely simplify things: AI tools are integrated as serverless functions i.e. they get called via a rest interface on the nuclio platform like some webservice. an image is sent from cvat to the service and it responds with the tracked location and state of the tracker. within this PR i have implemented such a service. from the perspective of implementation this is great, because of its simplicity - performance when tracking over multiple frames is not amazing with this approach.
> 
> to implement what you're looking for is not trivial. the simplest i can imagine is to call the tracking service repeatedly until the required amount of frames is tracked. BUT this is not very performant. each frame is sent in a separate http request to the service and it requires n requests to tracking n frames. a better solution would be to have a service that's capable of tracking multiple frames, keep its state internally and doesn't require the images to be sent in the request but rather access them from a docker mount. all of that would reduce flexibility but increase performance.

Thank you for the reply, my AI function also predict one frame at a time (batch size = 1), I can imagine the simplest is to repeatedly sent HTTP request, one after one automatically until all MP4 frames are sent. Thank you!

File: serverless/pytorch/dschoerk/transt/nuclio/main.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+def init_context(context):
+    context.logger.info("Init context...  0%")
+    model = create_tracker()
+    context.user_data.model = model
+    context.logger.info("Init context...100%")
+
+def log(msg):
+    #with open("/log.log", "a") as logf:
+    #    logf.write(msg+'\n')
+    pass
```
Comment (yasakova-anastasia) [inline serverless/pytorch/dschoerk/transt/nuclio/main.py:51]: I think this function can be removed.

File: serverless/pytorch/dschoerk/transt/nuclio/main.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
+import json
```
Comment (yasakova-anastasia) [inline serverless/pytorch/dschoerk/transt/nuclio/main.py:1]: Please add a license to the beginning of the file.
```
# Copyright (C) 2022 CVAT.ai Corporation
#
# SPDX-License-Identifier: MIT
```

File: serverless/pytorch/dschoerk/transt/nuclio/main.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        return context.Response(body=json.dumps(results), headers={},
+            content_type='application/json', status_code=200)
+
+    except Exception as e: # cavemen debugging
+        logf = open("/error.log", "w")
+        logf.write(str(e))
+        logf.write(traceback.format_exc())
+
+        return context.Response(headers={},
+            content_type='application/json', status_code=666)
```
Comment (yasakova-anastasia) [inline serverless/pytorch/dschoerk/transt/nuclio/main.py:144]: I think this should be removed.

File: serverless/pytorch/dschoerk/transt/nuclio/main.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    model = create_tracker()
+    context.user_data.model = model
+    context.logger.info("Init context...100%")
+
+def log(msg):
+    #with open("/log.log", "a") as logf:
+    #    logf.write(msg+'\n')
+    pass
+
+def encode_state(model):
```
Comment (yasakova-anastasia) [inline serverless/pytorch/dschoerk/transt/nuclio/main.py:53]: Could you please separate these functions into a separate `ModelHandler` class (as is done for other serverless functions)?

File: serverless/pytorch/dschoerk/transt/nuclio/main.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        context.logger.info("Run TransT model")
+        data = event.body
+        buf = io.BytesIO(base64.b64decode(data["image"]))
+        shapes = data.get("shapes")
+        states = data.get("states")
+
+        image = Image.open(buf).convert('RGB')
+        image = np.array(image)[:, :, ::-1].copy()
+
+        #cv2.imwrite('/test.jpg', image)
```
Comment (yasakova-anastasia) [inline serverless/pytorch/dschoerk/transt/nuclio/main.py:102]: Please remove all useless comments.

**Review by yasakova-anastasia — CHANGES_REQUESTED** [review]
@dschoerk,
Thanks for your contribution! I tested it, it works great. But I have some small comments. Could you fix them please?

**Comment by yasakova-anastasia** [comment]
I can't push to this thread. The requested URL returned error: 403. Opened another [PR](https://github.com/opencv/cvat/pull/5226).

**Comment by AIWithShrey** [comment]
> > Hi @dschoerk , thank for this really cool integration. This is very inspiring! Can I ask how you write the code for annotating video (MP4) in CVAT? I'm trying to implement a different task with endoscopy videos, I saw your video shows the CVAT processed each frame at a time until all frames are tracked/inference.
> > How are the images fed/transformed? In meantime, how are those requests kept posting to nutlio server until all frames(images) are done?
> > Thank you in advance and it would be super appreciated if there sample code for this video integration! Thanks!
> 
> at this time CVAT only supports to track one step at a time. a bounding box (seed) is drawn on the initially frame, and each time you press the "f" key to step to the next frame the objects are tracked. afaik there is no functionality at the moment to track multiple frames at once. in the mentioned video i just keep pressing the "f" key. i hope this answers your question.

How would you suggest I go by using TransT with multiple frames? Do I have to take the Docker route? I have about 1800 frames (30 fps, 1 minute long videos). Also, I don't find TransT on HuggingFace to use it on my own dataset, without having to use CVAT. Sure, CVAT makes my life easier for annotation, but it takes extremely long to annotate 1800 frames separately.

What would be the most feasible solution to my problem?

**Pull request closed without merging**